libraries/chain/include/eosio/chain/webassembly/eos-vm-oc.hpp

```diff
@@ -0,0 +1,794 @@
+#pragma once
+
+#include <eosio/chain/webassembly/common.hpp>
+#include <eosio/chain/exceptions.hpp>
+#include <eosio/chain/webassembly/runtime_interface.hpp>
+#include <eosio/chain/apply_context.hpp>
+#include <softfloat.hpp>
+#include "IR/Types.h"
+
+#include <eosio/chain/webassembly/eos-vm-oc/eos-vm-oc.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/memory.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/executor.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/code_cache.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/config.hpp>
+#include <eosio/chain/webassembly/eos-vm-oc/intrinsic.hpp>
+
+#include <boost/hana/equal.hpp>
+
+namespace eosio { namespace chain { namespace webassembly { namespace eosvmoc {
+
+using namespace IR;
+using namespace Runtime;
+using namespace fc;
+using namespace eosio::chain::webassembly::common;
+
+using namespace eosio::chain::eosvmoc;
+
+class eosvmoc_instantiated_module;
+
+class eosvmoc_runtime : public eosio::chain::wasm_runtime_interface {
+   public:
+      eosvmoc_runtime(const boost::filesystem::path data_dir, const eosvmoc::config& eosvmoc_config, const chainbase::database& db);
+      ~eosvmoc_runtime();
+      std::unique_ptr<wasm_instantiated_module_interface> instantiate_module(const char* code_bytes, size_t code_size, std::vector<uint8_t> initial_memory,
+                                                                             const digest_type& code_hash, const uint8_t& vm_type, const uint8_t& vm_version) override;
+
+      void immediately_exit_currently_running_module() override;
+
+      friend eosvmoc_instantiated_module;
+      eosvmoc::code_cache_sync cc;
+      eosvmoc::executor exec;
+      eosvmoc::memory mem;
+};
+
+/**
+ * class to represent an in-wasm-memory array
+ * it is a hint to the transcriber that the next parameter will
+ * be a size (in Ts) and that the pair are validated together
+ * This triggers the template specialization of intrinsic_invoker_impl
+ * @tparam T
+ */
+template<typename T>
+inline array_ptr<T> array_ptr_impl (size_t ptr, size_t length)
+{
+   constexpr int cb_full_linear_memory_start_segment_offset = OFFSET_OF_CONTROL_BLOCK_MEMBER(full_linear_memory_start);
+   constexpr int cb_first_invalid_memory_address_segment_offset = OFFSET_OF_CONTROL_BLOCK_MEMBER(first_invalid_memory_address);
+
+   static_assert(sizeof(T) == 64 || sizeof(T) == 32 || sizeof(T) == 20 || sizeof(T) == 16 || sizeof(T) == 8 ||
+                 sizeof(T) == 4 || sizeof(T) == 2 || sizeof(T) == 1, "array_ptr sizeof(T) only supports limited values");
+
+   if constexpr(sizeof(T) == 20) {
+      asm("shl $2, %[Length]\n"
+          "lea (%[Length], %[Length], 4), %[Length]\n"
```

Comment:
Is this really better then imul?  We're not running i386 here.  Actually, why is this being done in inline asm at all?

Reply:
that's the code clang generated for me for comparative c++. but agree, this probably shouldn't be asm.  it is holdover from an early impl that inspected overflow flags and such. the less inline asm the better